### ISSUE TITLE
How to generate tfrecord file with the ExampleListWithContext format?

### ISSUE THREAD

**Issue by fly12357** [tensorflow/ranking#141]

**Comment by xuanhuiwang** [comment]
We have an example in the test file here: https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_tfrecord_test.py#L111.

Let me know how this works for your case.

**Comment by fly12357** [comment]
> We have an example in the test file here: https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_tfrecord_test.py#L111.
> 
> Let me know how this works for your case.

Thanks for your help. I had generate my own tfrecord file with the ELWC format. But, when I feed th input feature_columns of my won data to the tfr.data.build_ranking_dataset. There were some bugs. 

The tfrecord generate code is :
                # Content
                content_features = tf.train.Features(feature=CONTENT_feature)
                content = tf.train.Example(features=content_features)

                ELWC = input_pb2.ExampleListWithContext()
                ELWC.context.CopyFrom(content)
                for example in EXAMPLES:
                    EXAMPLE = ELWC.examples.add()
                    EXAMPLE.CopyFrom(example)

                tf_serialized = ELWC.SerializeToString()

                writer.write(tf_serialized)

The  revised the tf_ranking_tfrecord.py  to suit my own data, the code are:

from absl import flags

import tensorflow as tf
import tensorflow_ranking as tfr

flags.DEFINE_enum(
    "data_format", "example_list_with_context",
    ["example_list_with_context", "example_in_example", "sequence_example"],
    "Data format defined in data.py.")
flags.DEFINE_string("train_path", "./TF_test.tfrecord", "Input file path used for training.")
flags.DEFINE_string("eval_path", None, "Input file path used for eval.")
flags.DEFINE_string("vocab_path", None,
                    "Vocabulary path for query and document tokens.")
flags.DEFINE_string("model_dir", None, "Output directory for models.")
flags.DEFINE_integer("batch_size", 32, "The batch size for train.")
flags.DEFINE_integer("num_train_steps", 15000, "Number of steps for train.")
flags.DEFINE_float("learning_rate", 0.05, "Learning rate for optimizer.")
flags.DEFINE_float("dropout_rate", 0.8, "The dropout rate before output layer.")
flags.DEFINE_list("hidden_layer_dims", ["64", "32", "16"],
                  "Sizes for hidden layers.")
flags.DEFINE_integer(
    "list_size", None,
    "List size used for training. Use None for dynamic list size.")
flags.DEFINE_integer("group_size", 1, "Group size used in score function.")
flags.DEFINE_string("loss", "approx_ndcg_loss",
                    "The RankingLossKey for the loss function.")
flags.DEFINE_string("weights_feature_name", "",
                    "The name of the feature where unbiased learning-to-rank "
                    "weights are stored.")
flags.DEFINE_bool("listwise_inference", False,
                  "If true, exports accept `data_format` while serving.")
flags.DEFINE_bool(
    "use_document_interactions", False,
    "If true, uses cross-document interactions to generate scores.")

FLAGS = flags.FLAGS

_LABEL_FEATURE = "label"
_PADDING_LABEL = -1
_EMBEDDING_DIMENSION = 20
_DOCUMENT_MASK = "__document_mask__"


def context_feature_columns():
    CONTENT_FC = {
        'L_bid_his': tf.io.FixedLenFeature(shape=(100), dtype=tf.float32),
        'L_inter_time': tf.io.FixedLenFeature(shape=(100), dtype=tf.float32),

        # 张量 3维
        'L_c1_his': tf.io.FixedLenFeature(shape=(), dtype=tf.string),
        'L_c2_his': tf.io.FixedLenFeature(shape=(), dtype=tf.string),
        'L_c3_his': tf.io.FixedLenFeature(shape=(), dtype=tf.string),
        'L_tag_his': tf.io.FixedLenFeature(shape=(), dtype=tf.string),
        # shape
        'L_c1_his_shape': tf.io.FixedLenFeature(shape=(2,), dtype=tf.int64),
        'L_c2_his_shape': tf.io.FixedLenFeature(shape=(2,), dtype=tf.int64),
        'L_c3_his_shape': tf.io.FixedLenFeature(shape=(2,), dtype=tf.int64),
        'L_tag_his_shape': tf.io.FixedLenFeature(shape=(2,), dtype=tf.int64)
    }
    return CONTENT_FC

def example_feature_columns():
    feature_columns = {
        # 标量
        'bid': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'bookWords': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'bookChapters': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'fans_5': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'fans_4': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'fans_3': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'fans_2': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'fans_1': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'cc_new': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'gc_new': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'bc_new': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'gc_ratio': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'Qfor72': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'Q_pr': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'zy_score': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'mean_sim': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'sum_cf': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'avg_cf': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'sum_cf_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'BKEMB_C1': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'BKEMB_C2': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'BKEMB_C3': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'BKEMB_C1_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'BKEMB_C2_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'BKEMB_C3_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'match_c1': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'match_c2': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'match_c3': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        'match_c1_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'match_c2_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        'match_c3_norm': tf.io.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),

        # 张量 2维
        'c1': tf.io.FixedLenFeature(shape=(2), dtype=tf.float32),
        'c2': tf.io.FixedLenFeature(shape=(4), dtype=tf.float32),
        'c3': tf.io.FixedLenFeature(shape=(4), dtype=tf.float32),
        'tag': tf.io.FixedLenFeature(shape=(19), dtype=tf.float32),
    }
    return feature_columns


def label_feature_column():
    DIC_LABEL = {
        'label': tf.compat.v1.FixedLenFeature(shape=(), dtype=tf.float32, default_value=None),
        }
    return DIC_LABEL

def make_input_fn(file_pattern,
                  batch_size,
                  randomize_input=True,
                  num_epochs=None):
  """Returns `Estimator` `input_fn` for TRAIN and EVAL.

  Args:
    file_pattern: (string) file pattern for the TFRecord input data.
    batch_size: (int) number of input examples to process per batch.
    randomize_input: (bool) if true, randomize input example order. It should
      almost always be true except for unittest/debug purposes.
    num_epochs: (int) Number of times the input dataset must be repeated. None
      to repeat the data indefinitely.

  Returns:
    An `input_fn` for `Estimator`.
  """
  #tf.compat.v1.logging.info("FLAGS.data_format={}".format(FLAGS.data_format))

  def _input_fn():
    """Defines the input_fn."""
    context_feature_spec = context_feature_columns()

    example_feature_spec = example_feature_columns()

    dataset = tfr.data.build_ranking_dataset(
        file_pattern=file_pattern,
        data_format="example_list_with_context",
        batch_size=3,
        list_size=None,
        context_feature_spec=context_feature_spec,
        example_feature_spec=example_feature_spec,
        reader=tf.data.TFRecordDataset,
        shuffle=randomize_input,
        num_epochs=num_epochs)
    features = tf.compat.v1.data.make_one_shot_iterator(dataset).get_next()

    return features

  return _input_fn


print(make_input_fn("./TF_test.tfrecord", None)())


The Error is 
![image](https://user-images.githubusercontent.com/40494145/69619058-ed7f9180-1075-11ea-8f7a-a401880f821a.png)

**Comment by fly12357** [comment]
> We have an example in the test file here: https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_tfrecord_test.py#L111.
> 
> Let me know how this works for your case.
Thanks for your help. I had generated my own tfrecord file with the ELWC format.

**Comment by xuanhuiwang** [comment]
I saw addition information in the email. For any feature, you need to set a valid default_value instead of None. This is because padding is needed to align to the `list_size1.

**Comment by fly12357** [comment]
> I saw addition information in the email. For any feature, you need to set a valid default_value instead of None. This is because padding is needed to align to the `list_size1.

Thanks so much. This problem has troubled me for a long time.

**Comment by xuanhuiwang** [comment]
Great! Closing this now.

**Comment by ratulray** [comment]
FWIW, we have done this at scale using Apache Beam which has inbuilt support for TFRecord format and protobuf
https://beam.apache.org/releases/javadoc/2.2.0/org/apache/beam/sdk/io/TFRecordIO.html
